evm/src/executor/abi.rs

```diff
@@ -680,6 +681,291 @@ pub static HARDHAT_CONSOLE_SELECTOR_PATCHES: Lazy<HashMap<Selector, Selector>> =
+        }
+    }
+}
+
+impl FormatValue for bool {
+    fn fmt(&self, spec: FormatSpec) -> String {
+        match spec {
+            FormatSpec::String => self.pretty(),
+            FormatSpec::Object => format!("'{}'", self.pretty()),
+            FormatSpec::Number => (*self as i32).to_string(),
+            FormatSpec::Integer => String::from("NaN"),
+        }
+    }
+}
+
+impl FormatValue for U256 {
+    fn fmt(&self, _spec: FormatSpec) -> String {
+        self.pretty()
+    }
+}
+
+impl FormatValue for I256 {
+    fn fmt(&self, _spec: FormatSpec) -> String {
+        self.pretty()
+    }
+}
+
+impl FormatValue for Address {
+    fn fmt(&self, spec: FormatSpec) -> String {
+        match spec {
+            FormatSpec::String => self.pretty(),
+            FormatSpec::Object => format!("'{}'", self.pretty()),
+            FormatSpec::Number | FormatSpec::Integer => String::from("NaN"),
+        }
+    }
+}
+
+impl FormatValue for Bytes {
+    fn fmt(&self, spec: FormatSpec) -> String {
+        match spec {
+            FormatSpec::String => self.pretty(),
+            FormatSpec::Object => format!("'{}'", self.pretty()),
+            FormatSpec::Number | FormatSpec::Integer => String::from("NaN"),
+        }
+    }
+}
+
+// Formats a `specstr` using the input values, v1, v2 and v3.
+// For example:
+//   console_log_format_n("%s has %d characters", 2, "foo", 3) == "foo has 3 characters"
+// num_v determines the maximum number of values to be used for the formatting (since we don't have
+// variadics in Rust). For example: console_log_format_n("%s", 1, v1, v2, v3) will format the string
+// using v1 only, ignoring v2 and v2.
+//
+// Formatting rules are the same as hardhat. The supported format specifiers are as follows:
+// - %s: Converts the value using its String representation. This is equivalent to applying
+//   UIfmt::pretty() on the format string.
+// - %d, %i: Converts the value to an integer. If a non-numeric value, such as String or Address, is
+//   passed, then the spec is formatted as `NaN`.
+// - %o: Treats the format value as a javascript "object" and converts it to its string
+//   representation.
+// - %%: This is parsed as a single percent sign ('%') without consuming any input value.
+//
+// Unformatted values are appended to the end of the formatted output using UIfmt::pretty().
+// If there are more format specifiers than values, then the remaining unparsed format specifiers
+// appended to the formatted output as-is.
+fn console_log_format_n(
+    specstr: &str,
+    num_v: isize,
+    v0: &dyn FormatValue,
+    v1: &dyn FormatValue,
+    v2: &dyn FormatValue,
+) -> String {
+    assert!(num_v <= 3);
+
+    let mut result = String::new();
+    let spec = specstr.as_bytes();
+    let mut expect_fmt = false;
+    let mut curr_value = 0;
+    for (pos, c) in spec.iter().enumerate() {
+        if curr_value >= num_v {
+            let suffix = String::from_utf8_lossy(&spec[pos..]);
```

Comment:
not sure I understand why we have to allocate a new string every iteration?

there is `String::chars()` which gives an iterator over the chars in the string, maybe this can be used here instead of iterating over the bytes?

Reply:
We only allocate once, since we break out of the loop immediately after.

The String conversion is so we can easily escape `%%`. It gets a bit complicated to do so using `specstr` because it's in utf-8: The `pos` index in `spec` may map to a different code point in `specstr`.